### ISSUE TITLE
Request: compute hash key in post step

### ISSUE THREAD

**Issue by wch** [actions/cache#135]
The cache action currently works well if the repository contains a file that represents the contents cached data -- for example, a `package-lock.json` file, which contains information on the exact versions of every package that's in the cache.

However, it does not work very well if the cache contents are not known ahead of time. Here's an example (note that this isn't exactly what we're trying to do, but it's similar). The purpose of a particular workflow is to test the repo using the latest versions of several external packages, so that we're alerted to problems early on. What I would like the workflow to do is:

1. restore the packages from the cache
2. update the packages to the latest versions
3. run tests
4. save the packages in a new cache object

This process doesn't work well with actions/cache, because the key is only determined at the beginning, in step 1, so the updated set of packages will never get saved. (This assumes they key uses a hash of the lockfile.)

What I would like to be able to do is compute the key in the post action. In other words hash the lockfile after the packages have been updated. 

I'm imagining something like this, where there's a `post_key` that's computed in the post cache step:

```yml
    post_key: npm-${{ hashFiles('package-lock.json') }}
    restore_keys: npm-
```

With this, it would always restore from the most recent accessible cache object, and save a new cache object only if any packages have been updated. That would be the ideal workflow for us. This would be really useful in any cases where the exact contents of the cache can't be known (or hashed) in the beginning.

Note: I know that this may not be the best way to work in JavaScript, but in the software ecosystem I work in (primarily with R), it is very common.

**Comment by dhadka** [comment]
Thanks for submitting this request!  It seems this is similar in nature to the discussion in #109.  The main difference is #109 is asking to always save the cache, whereas here we would only save it if the hash changes.  But the former could be accomplished by calculating the hash of all files included in the cache (although that may be computationally expensive).

Are there use cases where the `key` and `post_key` would use a different expression?  If we expect them to be the same, we can simplify this with something like:

```yaml
key: npm-${{ hashFiles('package-lock.json') }}
restore_keys: npm-
recalculate_key: true
```

**Comment by wch** [comment]
I like the simplicity of a boolean `recalculate_key` better than a separate `key` and `post_key`, but I think it would be problematic in practice.

Here's the problem case I can think of: Suppose that the file (or files) to be hashed isn't a part of the repository. Then you run into #106: if the files being hashed don't exist in the beginning, then `hashFiles()` errors out.

I've considered two possible solutions, but there are still problems. (Note that this discussion may be more relevant to #106.)
* If `hashFiles()` were changed so that, instead of erroring on missing files, it returned an empty string, then there's a risk of polluting your cache. If your workflow is misconfigured for at least one run, such the files to be hashed aren't created, but the workflow still successfully completes, then the cache post action will save a new object with a key like `npm-`, which would be an exact match for the `restore_keys`. This would cause problems later, because it alters the behavior in case of a cache miss -- if in a later run, the file to be hashed does exist, but `hashFiles()`returns a value that isn't yet in the cache, then instead of restoring the most recent cache object, this will cause it to restore the exact match, `npm-`, which contains no files.
* I think that if `hashFiles()` were changed so that, instead of erroring on missing files, it returned the _hash_ of an empty string (let's pretend it's `abcdef`), there's a similar problem. The good news is that you don't have to worry about accidentally saving an object for `npm-`. The bad news is that if you have one run where you accidentally save an object for `npm-abcdef`, then that empty cache object will be restored for every single run in the future, when what we really want is for it to restore from the most recent cache with the prefix `npm-`.

So in summary, if you use the same expression for the restore and save steps, it would be easy to accidentally save an object in the cache with so that the expression during the restore stage results in false positives. And once your cache is in that state, it's hard to correct it.

**Comment by mzabaluev** [comment]
Another use case for this is Rust library crates. By convention, library projects do not check `Cargo.lock` (the exact representation of the dependency state) into version control, unlike the projects that build end executables. But `Cargo.lock` is created or updated during every standalone library build (e.g. for test suite runs), so to improve performance of the dependency cache it's good to use the `Cargo.lock` checksum for the postfix of the cache key and use restore keys to match.

**Comment by mzabaluev** [comment]
For comparison, CircleCI has separate steps for saving and restoring the cache. This gives the developer full control as to when the cache is saved and how its key is calculated, and does not have the oddity of a post step not visible in the configuration.

Here's an example of how it could look like for the Rust library case above:

```yaml
    steps:
      - uses: actions/checkout@v1
      - name: Restore cargo registry cache
        id: cargo_cache
        uses: actions/restore-cache@v1
        with:
          path: ~/.cargo/registry
          # No exact key, uses restore-keys for an approx match
          restore-keys: |
            ${{ runner.os }}-cargo-registry-${{ hashFiles('**/Cargo.toml') }}-
            ${{ runner.os }}-cargo-registry-
      # Build and test, capturing dependency state in Cargo.lock
      # ...
      - name: Save cargo registry cache
        uses: actions/save-cache@v1
        with:
          path: ${{ steps.cargo_cache.outputs.path }}
          key: ${{ runner.os }}-cargo-registry-${{ hashFiles('**/Cargo.toml') }}-${{ hashFiles('Cargo.lock') }}
```

**Comment by ephemient** [comment]
⠀

**Comment by IceflowRE** [comment]
Iam also interested in this!
